app/models/service_terraform_template.rb

```diff
@@ -70,15 +73,37 @@ def check_completed(action)
   end
 
   def launch_terraform_template(action)
+    $embedded_terraform_log.debug("Service(#{id}).launch_terraform_template(#{action}) starts")
     terraform_template = terraform_template(action)
 
     # runs provision or retirement or reconfigure job, based on job_options
     stack = ManageIQ::Providers::EmbeddedTerraform::AutomationManager::Stack.create_stack(terraform_template, get_job_options(action))
+
+    # We save the newly created Job/OrchestrationStack.id, with job options, to identify the current running Job/OrchestrationStack.
+    # This especially is needed for Reconfigure action, as the action can run multiple times for same service instance, which creates multiple OrchestrationStack(job)s,
+    # unlike Provision or Retirement actions, each having single OrchestrationStack(job).
+    save_orchestration_stack_id(action, stack.id)
+
     add_resource!(stack, :name => action)
+
+    $embedded_terraform_log.debug("Service(#{id}).launch_terraform_template(#{action}) ends")
   end
 
   def stack(action)
-    service_resources.find_by(:name => action, :resource_type => 'OrchestrationStack').try(:resource)
+    case action
+    when ResourceAction::RECONFIGURE
+      # we need to reload, to pull updates to options from db,
+      # because the stack_id was saved in service instance in queue job
+      reload
+
+      job_options = options[job_option_key(action)]
+      orchestration_stack_id = job_options[:orchestration_stack_id]
+      $embedded_terraform_log.debug("find service_resource by resource_id:#{orchestration_stack_id}")
+
+      service_resources.find_by(:resource_id => orchestration_stack_id).try(:resource)
```

Comment:
If you're going to query on polymorphic object you have to include the `_type` as well, e.g. if there are 3 service_resources `[OrchestrationStack(id: 2), OrchestrationStack(id: 10), Vm(id: 2)]` then this won't always return the orchestration stack
```suggestion
      service_resources.find_by(:resource_type => 'OrchestrationStack', :resource_id => orchestration_stack_id).try(:resource)
```

Adding the action in there probably wouldn't hurt, I do wonder if we could use the same for provision/retire below as well

Reply:
Yes, we can use for provision/retire, as saving the orchestration_stack_id for all

Reply:
done